xdsl/transforms/common_subexpression_elimination.py

```diff
@@ -0,0 +1,185 @@
+from dataclasses import dataclass, field
+from typing import TypeVar
+
+from xdsl.dialects.builtin import ModuleOp, UnregisteredOp
+from xdsl.ir import Block, MLContext, Operation, Region, Use
+from xdsl.passes import ModulePass
+from xdsl.rewriter import Rewriter
+from xdsl.traits import IsolatedFromAbove, IsTerminator, Pure
+from xdsl.transforms.dead_code_elimination import is_trivially_dead
+
+
+@dataclass
+class OperationInfo:
+    """
+    Boilerplate helper to use in KnownOps cache.
+
+    This is to compare operations on their name, attributes, properties, results,
+    operands, and matching region structure.
+    """
+
+    op: Operation
+
+    def __hash__(self):
+        if isinstance(self.op, UnregisteredOp):
+            name = self.op.op_name.data
+        else:
+            name = self.op.name
+        return (
+            hash(name)
+            + sum(hash(i) for i in self.op.attributes.items())
+            + sum(hash(i) for i in self.op.properties.items())
+            + sum(hash(i.type) for i in self.op.results)
+            + sum(hash(i) for i in self.op.operands)
+        )
```

Comment:
May I recommend using `hash((a, b, c))` instead of `hash(a+b+c)` to combine hashes? This makes the concatenation non-symmetrical.

Reply:
Sure, good to know, thanks! That should help with performance if it ever comes down to that.
NB I kept the sum on the dicts because we want it to be symmetrical there.